<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `alignment` mod in crate `bio`."><meta name="keywords" content="rust, rustlang, rust-lang, alignment"><title>bio::alignment - Rust</title><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled ><script id="default-settings"></script><script src="../../storage.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../../favicon.svg">
<link rel="alternate icon" type="image/png" href="../../favicon-16x16.png">
<link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 8]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu">&#9776;</div><a href='../../bio/index.html'><div class='logo-container rust-logo'><img src='../../rust-logo.png' alt='logo'></div></a><p class="location">Module alignment</p><div class="sidebar-elems"><div class="block items"><ul><li><a href="#modules">Modules</a></li><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#types">Type Definitions</a></li></ul></div><p class="location"><a href="../index.html">bio</a></p><div id="sidebar-vars" data-name="alignment" data-ty="mod" data-relpath="../"></div><script defer src="../sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu"><img src="../../brush.svg" width="18" alt="Pick another theme!"></button><div id="theme-choices" role="menu"></div></div><script src="../../theme.js"></script><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" class="help-button">?</button>
                <a id="settings-menu" href="../../settings.html"><img src="../../wheel.svg" width="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Module <a href="../index.html">bio</a>::<wbr><a class="mod" href="">alignment</a></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../../src/bio/alignment/mod.rs.html#6-14" title="goto source code">[src]</a></span></h1><div class="docblock"><p>Various alignment and distance computing algorithms.</p>
</div><h2 id="modules" class="section-header"><a href="#modules">Modules</a></h2>
<table><tr class="module-item"><td><a class="mod" href="distance/index.html" title="bio::alignment::distance mod">distance</a></td><td class="docblock-short"><p>Various subroutines for computing a distance between sequences. Features
both scalar and efficient vectorized distance functions with SIMD.</p>
</td></tr><tr class="module-item"><td><a class="mod" href="pairwise/index.html" title="bio::alignment::pairwise mod">pairwise</a></td><td class="docblock-short"><p>Calculate alignments with a generalized variant of the Smith Waterman algorithm.
Complexity: O(n * m) for strings of length m and n.</p>
</td></tr><tr class="module-item"><td><a class="mod" href="poa/index.html" title="bio::alignment::poa mod">poa</a></td><td class="docblock-short"><p>Partial-Order Alignment for fast alignment and consensus of multiple homologous sequences.</p>
</td></tr><tr class="module-item"><td><a class="mod" href="sparse/index.html" title="bio::alignment::sparse mod">sparse</a></td><td class="docblock-short"><p>Calculate 'sparse' alignments from kmer matches. Can be much faster than
Smith-Waterman for long string, when a large enough k is used.
Complexity: O(n * log(n)) for a pair of strings with n k-kmer matches. This
approach is useful for generating an approximate 'backbone' alignments
between two long sequences, for example in long-read alignment or
genome-genome alignment. The backbone alignment can be used as-is, or can serve
as a guide for a banded alignment.  By tuning k so that len(query) + len(reference) &lt; 4^k,
the number of false positive kmer matches is kept small, resulting in very
fast run times for long strings.</p>
</td></tr></table><h2 id="structs" class="section-header"><a href="#structs">Structs</a></h2>
<table><tr class="module-item"><td><a class="struct" href="struct.Alignment.html" title="bio::alignment::Alignment struct">Alignment</a></td><td class="docblock-short"><p>We consider alignment between two sequences x and  y. x is the query or read sequence
and y is the reference or template sequence. An alignment, consisting of a score,
the start and end position of the alignment on sequence x and sequence y, the
lengths of sequences x and y, and the alignment edit operations. The start position
and end position of the alignment does not include the clipped regions. The length
of clipped regions are already encapsulated in the Alignment Operation.</p>
</td></tr></table><h2 id="enums" class="section-header"><a href="#enums">Enums</a></h2>
<table><tr class="module-item"><td><a class="enum" href="enum.AlignmentMode.html" title="bio::alignment::AlignmentMode enum">AlignmentMode</a></td><td class="docblock-short"><p>The modes of alignment supported by the aligner include standard modes such as
Global, Semi-Global and Local alignment. In addition to this, user can also invoke
the custom mode. In the custom mode, users can explicitly specify the clipping penalties
for prefix and suffix of strings 'x' and 'y' independently. Under the hood the standard
modes are implemented as special cases of the custom mode with the clipping penalties
appropriately set.</p>
</td></tr><tr class="module-item"><td><a class="enum" href="enum.AlignmentOperation.html" title="bio::alignment::AlignmentOperation enum">AlignmentOperation</a></td><td class="docblock-short"><p>Alignment operations supported are match, substitution, insertion, deletion
and clipping. Clipping is a special boundary condition where you are allowed
to clip off the beginning/end of the sequence for a fixed clip penalty. The
clip penalty could be different for the two sequences x and y, and the
clipping operations on both are distinguishable (Xclip and Yclip). The usize
value associated with the clipping operations are the lengths clipped. In case
of standard modes like Global, Semi-Global and Local alignment, the clip operations
are filtered out</p>
</td></tr></table><h2 id="types" class="section-header"><a href="#types">Type Definitions</a></h2>
<table><tr class="module-item"><td><a class="type" href="type.TextSlice.html" title="bio::alignment::TextSlice type">TextSlice</a></td><td class="docblock-short"></td></tr></table></section><section id="search" class="content hidden"></section><section class="footer"></section><div id="rustdoc-vars" data-root-path="../../" data-current-crate="bio"></div>
    <script src="../../main.js"></script><script defer src="../../search-index.js"></script></body></html>